### ISSUE TITLE
Declaring additional foreign key constraints

### ISSUE THREAD

**Issue by tysonzero** [yesodweb/persistent#1050]
Currently foreign keys are only created via directly using a `###Id` type in another column. However there are a variety of use cases that fit outside of this.

For example it is useful to have the primary key be a foreign key to the parent type, in order to implement table per class inheritance.

In general SQL allows foreign keys to go from any set of columns in one table to an equally typed set of columns in another table.

It would great to be able to declare these foreign keys in Persistent.

One interesting thing to consider if that in SQL both ends of foreign key have to have identical types (`uuid`, `int8` etc.). However in Persistent it would be useful to allow for a super-type foreign key, such as `DogId(uuid)` -> `AnimalId(uuid)`. It would also be useful if those foreign keys came with an "upcast" instance to allow for easy and safe conversions.

**Comment by tysonzero** [comment]
I have discovered that you can directly create foreign keys via `Foreign` syntax.

I can't find any documentation for it anywhere but from source diving I think I know the current capabilities.

The main missing functionality I can see is:

* multi-column foreign keys

* foreign keys between different Haskell types that share a single DB type

**Comment by parsonsmatt** [comment]
Multicolumn foreign keys should be supported. You have to list the columns in the current table (in the same order) that line up with the fields in the primary key in the target table. From the tests:

```
  ParentComposite
    name String
    lastName String
    Primary name lastName

  ChildComposite
    pname String
    plastName String
    Foreign ParentComposite OnDeleteCascade fkparent pname plastName
    deriving Show Eq
```

This does need to be documented.

**Comment by tysonzero** [comment]
Ah ok gotcha.

I generally only use single-column primary keys, so when I'm using multi-column foreign keys they point to non-primary fields.

**Comment by parsonsmatt** [comment]
Yeah, we have another issue for foreign key to point to non-primary key (#957) that we need to get fixed.
